### PR TITLE
Move to circleci

### LINKED ISSUE #160: Move away from travisCI
travis service is not as good as it was for open source projects.
Alternatives which come to my mind currently:

 - circle CI
 - github actions

@dylanhitt What do you think? Any ideas on this?

### PR DESCRIPTION
**Pull request by SimonBaeumer**
Fixes #160 

- macOS build is disabled, on circle CI it needs a subscription
- go integration tests now run in docker with `./test.sh` or using its make target `make test-coverage-all-dockerized`
   -  this was necessary because circle CI does not support docker port-forwarding
   - decided against docker-compose because the test container would directly exit or needs to wait

## Checklist

 - [x] Added unit / integration tests for windows, macOS and Linux? 
 - [x] Added a changelog entry in [CHANGELOG.md](https://github.com/commander-cli/commander/blob/master/CHANGELOG.md)?
 - [x] Updated the documentation ([README.md](https://github.com/commander-cli/commander/blob/master/README.md), [docs](https://github.com/commander-cli/commander/blob/master/docs))?
 - [x] Does your change work on `Linux`, `Windows` and `macOS`?

### CODE REVIEW

**Comment by SimonBaeumer** [comment]
@dylanhitt would like to have your review on this. windows builds are capped, but I think it is acceptable given this is a small open source project 👍

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          version: 19.03.13
+      - run:
+          name: Install commander
+          command: |
+            mkdir -p ~/.local/bin/
+            curl -L https://github.com/commander-cli/commander/releases/download/v2.1.0/commander-linux-amd64 -o ~/.local/bin/commander
+            chmod +x ~/.local/bin/commander
+      - run: make integration-linux-dockerized
+
+  integration-windows:
```
Comment (dylanhitt) [inline .circleci/config.yml:67]: We could guarantee some coverage by using `powershell` on linux. We'd be losing the ability to test `cmd`. How often are we able to run windows tests with our current support tier? I see their free tier states `2,500 free credits/week`, but I can't find any information stating that windows is capped?

File: test.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+docker rm commander-integration-go-test || true
+docker network rm commander_test || true
+
+docker build -t commander-int-ssh-server -f integration/containers/ssh/Dockerfile .
+docker network create  --driver=bridge --subnet=172.28.0.0/16 commander_test
+
+docker run -d \
+  --ip=172.28.0.2 \
+  --network commander_test \
+  --name int-ssh-server \
```
Comment (dylanhitt) [inline test.sh:24]: I think you accidentally added two names here.

File: test.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  --name commander-int-ssh-server \
+  commander-int-ssh-server
+
+docker build -t commander-int-test -f integration/containers/test/Dockerfile .
+docker run \
```
Comment (dylanhitt) [inline test.sh:29]: you can always add `docker run --rm` here. Avoiding some of the code below. But, you probably have a valid reason for not doing so

File: Makefile [hunk 4]
```diff
@@ -1,90 +1,120 @@
...
 	commander test commander_unix.yaml
 
+.PHONY: integration-linux
 integration-linux: build
 	$(info INFO: Starting build $@)
-	./integration/setup_unix.sh
-	commander test commander_unix.yaml
-	commander test commander_linux.yaml
-	./integration/teardown_unix.sh
+	DOCKER_HOST=${DOCKER_HOST} DOCKER_CERT_PATH=${DOCKER_CERT_PATH} commander test commander_unix.yaml
```
Comment (dylanhitt) [inline Makefile:79]: I can't quite figure out where these vars are coming from?

**Review by dylanhitt — COMMENTED** [review]
This looks good, I got to learn quite a bit from this 😄 . I'm not concerned about not being able to test `macOS`. If we run into a problem with it later we can discuss some possible solutions. 

You mentioned windows being capped? I can't really find that in their docs. If so how may windows builds do we get per X time frame?

File: .circleci/config.yml (hunk 1 above)
Comment (SimonBaeumer) [reply .circleci/config.yml:67]: Do we get the same outputs from it as on windows, i.e. path separators?
In general we don't test the shell, it is about the operating system commander runs on.

File: test.sh (hunk 2 above)
Comment (SimonBaeumer) [reply test.sh:24]: Good catch, thanks!

File: test.sh (hunk 3 above)
Comment (SimonBaeumer) [reply test.sh:29]: 👍  Just forgot it :D
Added `--rm`, time shows how this container evolves. If it is used more often it definitely needs to change and cached.

File: Makefile (hunk 4 above)
Comment (SimonBaeumer) [reply Makefile:79]: They are coming from circle CI and are default docker env variables. 
If docker runs on a remote host it is necessary to provide these paths.

**Comment by SimonBaeumer** [comment]
I am not sure as well how much windows test capacity circleci provides.
Given the builds would be broken for an entire week because the credits are exhausted the windows tests should only be executed on master branch and release tags.

File: .circleci/config.yml (hunk 1 above)
Comment (dylanhitt) [reply .circleci/config.yml:67]: I believe so, but i agree. I don't think it's really worth investigating

File: test.sh (hunk 3 above)
Comment (dylanhitt) [reply test.sh:29]: You can remove lines, 40 and 41 as well

**Comment by dylanhitt** [comment]
> Given the builds would be broken for an entire week because the credits are exhausted the windows tests should only be executed on master branch and release tags.

We could possibly run into some issues if we end up in a tight debugging spot. I have a laptop which I break out occasionally for windows problems, but maybe we should think of a branching strategy if we run into an issue on windows on the master branch. Say open up a branch called `win-<fixname>` that will run the windows tests? Or we can always just address them before a release. 

I think I'm personally making an issue out of nothing. Just my two cents. I'd be fine with just running windows test on master/tag. Once that's done I'd be happy to approve.

**Comment by dylanhitt** [comment]
Actually looking into it more, it appears we're capped out completely. The 2500 credits are for both linux/windows. The windows jobs are taking a [little under 200 credits per run..](https://circleci.com/pricing/). So not testing windows on every branch would be fine. Maybe we can still test on windows [circle does have the ability to only build on pull request](https://circleci.com/docs/2.0/oss/#only-build-pull-requests). But then we wouldn't be building on the branch 🤷. 

Honestly, I'm fine with just running the docker based builds until master/tags. We could possibly add some type of branching exception if we need it for debugging like `win-<name>`. I could also always break out my windows machine 🙃.

File: test.sh (hunk 3 above)
Comment (SimonBaeumer) [reply test.sh:29]: Right, done

**Comment by SimonBaeumer** [comment]
@dylanhitt looks better on the linux side:
```

Organizations on our free plan get 400,000 free credits per month ($2,880 annual value) to use for Linux open source builds. Enabling this will allow this project’s builds to use them and let others see your builds, both through the web UI and the API.
```

Therefore we can execute windows build just on master & tags that we do not release broken versions for windows.
Enabled your suggestion to only build on PRs, which is okay imho, if I do not create a pull request it is not that of a good PR anyways :D

**Review by dylanhitt — APPROVED** [review]
This is awesome, again let's see how it goes.

**Pull request merged**